src/index.ts

```diff
@@ -1,67 +1,146 @@
-import {ILayoutRestorer, JupyterFrontEnd, JupyterFrontEndPlugin} from '@jupyterlab/application';
-import '../style/index.css';
-import {GlobusHome} from "./globus/home";
-import {IDocumentManager} from '@jupyterlab/docmanager';
-import {IFileBrowserFactory} from "@jupyterlab/filebrowser";
-import {GlobusWidgetManager} from "./globus/widget_manager";
+import { buildIcon, reactIcon } from '@jupyterlab/ui-components';
+import { IDocumentManager } from '@jupyterlab/docmanager';
+import { IFileBrowserFactory } from '@jupyterlab/filebrowser';
+import { ILayoutRestorer, JupyterFrontEnd, JupyterFrontEndPlugin } from '@jupyterlab/application';
+import { MainAreaWidget } from '@jupyterlab/apputils';
+import { PageConfig } from '@jupyterlab/coreutils';
 
+import { GlobusHome } from './globus/home';
+import { GlobusWidgetManager } from './globus/widget_manager';
+import { GlobusWidget } from './widget';
 import { requestAPI } from './handler';
+
+import '../style/index.css';
+
+const addJupyterCommands = (app: JupyterFrontEnd, factory: IFileBrowserFactory, commands: Array<any>) => {
+  for (let command of commands) {
+    app.commands.addCommand(command.command, {
+      label: command.label,
+      caption: command.caption,
+      icon: buildIcon,
+      execute: async () => {
+        var files = factory.tracker.currentWidget.selectedItems();
+        var jupyterToken = PageConfig.getToken();
+        var label =
+          command.transfer_direction == 'to_collection'
+            ? 'Transfer to a Globus Collection'
+            : 'Transfer from a Globus Collection';
+
+        let fileArray = [],
+          fileCheck = true;
+        while (fileCheck) {
+          let file = files.next();
+          if (file) {
+            fileArray.push(file);
+          } else {
+            fileCheck = false;
+          }
+        }
+
+        const config = await requestAPI<any>('config');
+        const isAuthenticated = config.is_logged_in;
+
+        if (isAuthenticated) {
+          const content = new GlobusWidget(jupyterToken, fileArray, command.transfer_direction);
+          const widget = new MainAreaWidget<GlobusWidget>({ content });
+          widget.title.label = label;
+          widget.title.icon = reactIcon;
+          app.shell.add(widget, 'main');
+        } else {
+          let authInterval = window.setInterval(async () => {
+            const config = await requestAPI<any>('config');
+            const isAuthenticated = config.is_logged_in;
+
+            const content = new GlobusWidget(jupyterToken, fileArray, command.transfer_direction);
+            const widget = new MainAreaWidget<GlobusWidget>({ content });
+            widget.title.label = label;
+            widget.title.icon = reactIcon;
+
+            if (isAuthenticated) {
+              app.shell.add(widget, 'main');
+              clearInterval(authInterval);
+            }
+          }, 500);
+          window.open('/globus-jupyterlab/login', '_blank');
```

Comment:
I think this will open a new tab for the user to login, right? 

Another alternative might be opening a small separate window for them to login:

`window.open(url, 'Globus Login', 'height=600,width=800').focus();`

That way the jupyter page never leaves the user's sight.

Reply:
I think I could also use a bit more clarification, does the `else` block above wait for the user to login?

Reply:
1. Yes the '_blank' setting opens the window in a new tab. Good idea about opening in a window so the user can still view the Jupyter notebook.
2. If the user is logged in, the react widget will mount immediately; if not, the setTimeout waits for the user to login before adding the widget to the shell.